### PR TITLE
Fail the query with a helpful message if the variables defined do not match the expected values

### LINKED ISSUE #3855: [graphql] Error running query: argument of type 'NoneType' is not iterable
**Describe the bug**

Trying a new query to retrieve the offline data, used in the Cylc UI (for the offline UI, https://github.com/cylc/cylc-ui/pull/505).

```
{
  "errors": [
    {
      "message": "argument of type 'NoneType' is not iterable",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "workflows"
      ]
    }
  ],
  "data": {
    "workflows": null
  }
}
```

**Release version(s) and/or repository branch(es) affected?**

`master` for UIS (01988bb66307e53a1bda66e9ead70c64e0cda9c9) and Flow (bd293959edcf6db96414d608ef43772299514e36)

**Steps to reproduce the bug**

Run the query in GraphiQL, with a workflow on hold.

**Expected behavior**

Valid data retrieved, or an error response with the explanation.

**Screenshots**

**Additional context**

Query

```
query ($workflowID: ID) {
  workflows (ids: [$workflowID]) {
    ...WorkflowData
    cyclePoints: familyProxies (ids: ["root"], ghosts: true) {
      ...CyclePointData
    }
    taskProxies (sort: { keys: ["name"], reverse: false }, ghosts: true) {
      ...TaskProxyData
      jobs(sort: { keys: ["submit_num"], reverse:true }) {
        ...JobData
      }
    }
    familyProxies (exids: ["root"], sort: { keys: ["name"] }, ghosts: true) {
      ...FamilyProxyData
    }
  }
}

fragment WorkflowData on Workflow {
  id
  name
  status
  owner
  host
  port
}

fragment CyclePointData on FamilyProxy {
  id
  cyclePoint
}

fragment FamilyProxyData on FamilyProxy {
  id
  name
  state
  cyclePoint
  firstParent {
    id
    name
    cyclePoint
    state
  }
}

fragment TaskProxyData on TaskProxy {
  id
  name
  state
  isHeld
  cyclePoint
  latestMessage
  firstParent {
    id
    name
    cyclePoint
    state
  }
  task {
    meanElapsedTime
    name
  }
}

fragment JobData on Job {
  id
  firstParent: taskProxy {
    id
  }
  batchSysName
  batchSysJobId
  host
  startedTime
  submittedTime
  finishedTime
  state
  submitNum
}
```

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

### PR DESCRIPTION
**Pull request by kinow**
These changes close #3855 

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.
- [x] No dependency changes.

### CODE REVIEW

**Comment by kinow** [comment]
Now if I accidentally misspell a variable in Cylc UI, the message should hint that there's something wrong with the variables.

image

And fixing the variables then the query runs just fine.

image

**Comment by kinow** [comment]
Virtual memory error when running the bash tests in the Docker container. Fix available in #3854 

Will spend some more time on this later trying to add a test. Not sure if a unit test will work, might try our new test classes for the first time :+1:

**Comment by kinow** [comment]
The other PR that contains the fix for the Docker builds may take a while to be merged, while we wait for GH Actions to include Python 3.9.

So I've cherry-picked the commit that should fix Docker, and pushed here :+1:

**Comment by kinow** [comment]
Functional test broadcast/00-simple is failing, but I think it needs to be updated after this fix (just need to understand how it works, and confirm it indeed just needs to be updated)

File: cylc/flow/scripts/cylc_broadcast.py
Comment (kinow) [inline cylc/flow/scripts/cylc_broadcast.py:265]: @dwsutherland I had a functional test (`broadcast/00-simple.t`) failing as `bCutoff` (which is created using this `expire` value) was set to `None`.

In `cylc-flow`, in `network/graphql.py`, the `get_variable_values` function removes variables that are set to `None`. Which results in a GraphQL without all the variables defined.

If you run a query in JavaScript, or in GraphiQL and the number of variables in the query definition does not match the number of variables provided in the client query, it results in an error.

Apparently it was working for this functional test, but I think it's safe to set the default here to an empty string? The tests pass fine. WDYT?

**Comment by kinow** [comment]
I fixed the functional test in this PR, and yesterday it passed in the Docker image. Strangely, it failed now, but my only updates were changelog and a new **unit** test. :thinking: ready for review, I think the test is just a bit brittle in Docker (will investigate why later)

**Review by hjoliver — APPROVED** [review]
LGTM, but I'll defer to @dwsutherland for the definitive review.

**Review by dwsutherland — CHANGES_REQUESTED** [review]
Good idea @kinow , may I suggest the following instead:
```
        for defn in document_ast.definitions:
            if isinstance(defn, ast.OperationDefinition):
                root_type = get_operation_root_type(schema, defn)
                definition_variables = defn.variable_definitions or []
                if definition_variables:
                    def_var_names = {
                        v.variable.name.value
                        for v in definition_variables
                    }
                    var_names_diff = def_var_names.difference({
                        k
                        for k in variable_values
                        if k in def_var_names
                    })
                    # check if we are missing some of the definition variables
                    if var_names_diff:
                        msg = (f'Please check your query variables. The '
                               f'following variables are missing: '
                               f'[{", ".join(var_names_diff)}]')
                        raise ValueError(msg)
                self.operation_defs[getattr(defn.name, 'value', root_type)] = {
                    'definition': defn,
                    'parent_type': root_type,
                    'variables': get_variable_values(
                        schema,
                        definition_variables,
                        variable_values
                    ),
                }
            elif isinstance(defn, ast.FragmentDefinition):
                self.fragment_defs[defn.name.value] = defn
```

Then you don't have to change the default behaviour of GraphQL (i.e. if `None` is desirable to fall to the schema defined default), and hence not have to change the CLI tests.

**Comment by kinow** [comment]
> Good idea @kinow , may I suggest the following instead:
> 
> Then you don't have to change the default behaviour of GraphQL (i.e. if `None` is desirable to fall to the schema defined default), and hence not have to change the CLI tests.

Hi @dwsutherland ! Thanks for the suggestion! I will just need some time tomorrow to re-contextualize with the change, and understand your change. Then it should be a matter of just updating the PR with a new/edit commit, and check if tests pass :+1: 

Might be able to complete it before the executive roadshow starts tomorrow in Auckland, by the morning.

**Comment by kinow** [comment]
Fixed git conflicts, and applied @dwsutherland 's patch (thanks!). I think we now just need to wait for CI to verify.

File: cylc/flow/scripts/cylc_broadcast.py
Comment (dwsutherland) [reply cylc/flow/scripts/cylc_broadcast.py:265]: Probably want to change this back now?

Comment (kinow) [reply cylc/flow/scripts/cylc_broadcast.py]: I've dropped the commit with that change, and push-force'd. Now it's up to GitHub actions to confirm everything's working :+1:

**Review by dwsutherland — APPROVED** [review]
Good to go 👍

**Pull request merged**
